src/hipscat_import/run_import.py

```diff
@@ -1,54 +1,76 @@
 """Import a set of non-hipscat files using dask for parallelization"""
 
+
 import hipscat.io.write_metadata as io
 import numpy as np
 from dask.distributed import Client, as_completed
 from hipscat import pixel_math
 from tqdm import tqdm
 
 import hipscat_import.map_reduce as mr
+import hipscat_import.resume_files as rf
 from hipscat_import.arguments import ImportArguments
 
 
 def _map_pixels(args, client):
     """Generate a raw histogram of object counts in each healpix pixel"""
 
+    raw_histogram = rf.read_histogram(args.tmp_path, args.highest_healpix_order)
+    if rf.is_mapping_done(args.tmp_path):
+        return raw_histogram
+
+    mapped_paths = rf.read_mapping_keys(args.tmp_path)
+
     futures = []
     for i, file_path in enumerate(args.input_paths):
+        if file_path in mapped_paths:
+            continue
         futures.append(
             client.submit(
                 mr.map_to_pixels,
+                key=file_path,
                 input_file=file_path,
                 file_reader=args.file_reader,
                 filter_function=args.filter_function,
                 highest_order=args.highest_healpix_order,
                 ra_column=args.ra_column,
                 dec_column=args.dec_column,
                 shard_suffix=i,
-                cache_path=None if args.debug_stats_only else args.tmp_dir,
+                cache_path=None if args.debug_stats_only else args.tmp_path,
             )
         )
 
-    raw_histogram = pixel_math.empty_histogram(args.highest_healpix_order)
-    for _, result in tqdm(
+    for future, result in tqdm(
         as_completed(futures, with_results=True),
         desc="Mapping  ",
         total=len(futures),
         disable=(not args.progress_bar),
     ):
         raw_histogram = np.add(raw_histogram, result)
+        rf.write_histogram(args.tmp_path, raw_histogram)
+        rf.write_mapping_key(args.tmp_path, future.key)
+    rf.set_mapping_done(args.tmp_path)
```

Comment:
What happens if the job crashes between lines 51 and 52? The histograms and mapping keys have been written out, but the mapping file handle hasn't been set. I realize this is a very low likelihood probability, but thought I would ask. Would this end up reprocessing those entries and double counting?

Reply:
I think if the mapping key has been written then the next time it is run that file would be skipped from processing in lines 26 and 27

Reply:
Yeah - I don't know of any clever way to make these two operations atomic. I think we could even get killed mid-write and only part of the line would be available (or some subset of the histogram file would be updated), so combining into a single file doesn't resolve the problem.

This *would* cause a failure in the *reduce* stage, in a weird way:
- say input 5 has 12 objects in pixel 7
- if the job is killed between these lines, the histogram will have 12 objects in pixel 7, but the mapping key won't be present
- on resume, we'll still process input 5. we'll only overwrite the shard files generated on the first pass, and not write *new* ones.
- we'll add another 12 objects to pixel 7
- when we reduce, we set an expectation of finding 24 objects at pixel 7, but only 12 are written
- we raise an error inside reduce pixels

If we switch the order of these two lines and the job is killed between the steps:
- the histogram will have 0 objects in pixel 7, and the mapping key will be present
- on resume, we won't process input 5, so the histogram will still have 0 objects in pixel 7.
- when we reduce, we'll find 12 objects, expecting 0, and raise an error.

I'm open to suggestions!

Reply:
Warning: I don't have a **good** suggestion. I was wondering if you should include another file-based indicator (called start) that you write before the writes begin. When you are resuming you can skip anything that has "start" and "done", do anything that has neither, and fail if it only has one (or delete the intermediate files and restart).

Reply:
To be clear that wouldn't cover every case. I also don't have a way to make it fully atomic.

Reply:
All this might be over-engineering for a very rare case though.

Reply:
@delucchi-cmu is this is a problem about the operations on line 50 and 51 (Writing the histogram and writing the making key) or on line 51 and 52 (writing the histogram and keys and writing the done file)?

Reply:
If it's the former, another start writing log file as Jeremy suggests could work, if it's the latter, I'm not sure I understand the issue, sorry.

Reply:
Ah! I misunderstood the original question. If we fail between 51 and 52, then that's totally fine. When we pick back up to resume
- we'll load the `mapping_histogram`, which will contain the full histogram at that point
- we'll load the `mapping_log`, which will contain all the input files
- as we loop through the input files, we'll just skip all of them
- at the end of the loop, we'll write out the done file
- if we need to resume again after that, the `mapping_done` file will be around and everything will be peachy.

Adding extra guard around the histogram writing so we can find out quickly that the intermediate files have been corrupted.